FRR: T4020: Added CLI options for FRR daemons

zdc opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change Summary
<!--- Provide a general summary of your changes in the Title above -->
Added CLI options for FRR daemons

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- All submitted PRs must be linked to a Task on Phabricator. -->
* https://phabricator.vyos.net/T4020

## Component(s) name
<!-- A rather incomplete list of components: ethernet, wireguard, bgp, mpls, ldp, l2tp, dhcp ... -->
FRR

## Proposed changes
<!--- Describe your changes in detail -->
Added first CLI items for controlling FRR daemons parameters that cannot be changed via vtysh and are available via arguments only. Now it is possible to enable/disable modules: SNMP (for each daemon), BMP (for BGP), IRDP (for Zebra).

## How to test
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```
-->
There is a new section in CLI which can be used to change the `/etc/frr/daemons` file. For example:
```
set system frr bmp
set system frr irdp
set system frr snmp bgpd
set system frr snmp zebra
```
should enable SNMP and IRDP modules for zebra, and SNMP and BMP for  bgpd:
```
zebra_options="  -s 90000000 --daemon -A 127.0.0.1 -M snmp -M irdp"
bgpd_options="   --daemon -A 127.0.0.1 -M rpki -M snmp -M bmp"
```


## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

zdc committed:
FRR: T4020: Added CLI options for FRR daemons

Added first CLI items for controlling FRR daemons parameters that
cannot be changed via vtysh and are available via arguments only. Now
it is possible to enable/disable modules: SNMP (for each daemon),
BMP (for BGP), IRDP (for Zebra).

c-po commented on src/conf_mode/system_frr.py:
```diff
@@ -0,0 +1,173 @@
+
+def generate(frr_config):
+    # read daemons config file
+    daemons_config = read_file(config_file)
```
Please do not read/modify/write the daemon file. We always render a full template which gives less possibilities of errors. Also that is the "standard" way.

zdc replied:
Sure, this makes the task even easier. Done.

c-po commented on src/conf_mode/snmp.py:
```diff
@@ -581,7 +581,11 @@ def apply(snmp):
-    call('vtysh -c "configure terminal" -c "agentx" >/dev/null')
+    # This should be done for each daemon individually because common command
+    # works only if all the daemons started with SNMP support
+    frr_daemons_list = ['bgpd', 'ospf6d', 'ospfd', 'ripd', 'zebra']
```
What is the benefit of enabling this only for selected daemons instead of all?

c-po replied:
If it should be enabled for a single daemon, we should read in the system rrr settings?

zdc replied:
There is one nuance with `vtysh` in this case: if not all daemons have the snmp module loaded (what is possible now) the command will not be applied to any of them. That is why we need to execute the command for each daemon individually.
This is some kind of workaround to keep SNMP settings logic simple and not involve using newly added CLI options there, which still can be the case after we will see that everything works as expected.

c-po commented on src/conf_mode/system_frr.py:
```diff
@@ -0,0 +1,173 @@
+        conf = Config()
+    base = ['system', 'frr']
+    if not conf.exists(base):
+        return {}
```
`conf.get_config_dict(base)` returns an empty dict if not found - please drop the superfluous code

zdc replied:
Thanks, done!

zdc committed:
FRR: T4020: Updated CLI options processing for FRR daemons

Instead of analyzing options for each daemon now we use a single
template for the whole configuration file. This makes logic a bit less
flexible, but much easier.
Removed unnecessary check for returned by the
`conf.get_config_dict(base)` config.

Also, added the ability to disable `strip()` of file content while
using `read_file()` what is necessary for proper comparing with
updated content.

c-po commented on data/templates/frr/daemons.frr.tmpl:
```diff
@@ -0,0 +1,50 @@
+ripd_options="   --daemon -A 127.0.0.1
+{%- if snmp is defined and snmp.ripd is defined %} -M snmp{% endif -%}
+"
+ripngd_options=" --daemon -A ::1"
```
Why not also add SNMP support for RIPng, ISIS and LDP?

zdc replied:
I see that the SNMP module was added for more daemons lately, maybe with one of the FRR updates. Updated the list to contain all the daemons available now in 1.4.

c-po commented on interface-definitions/system-frr.xml.in:
```diff
@@ -0,0 +1,66 @@
+        <children>
+          <leafNode name="bmp">
+            <properties>
+              <help>>Enable BGP Monitoring Protocol support</help>
```
What is this additional `>`?

zdc replied:
This was a typo. Fixed.

c-po commented on python/vyos/util.py:
```diff
@@ -182,16 +182,20 @@ def call(command, flag='', shell=None, input=None, timeout=None, env=None,
 
 
-def read_file(fname, defaultonfailure=None):
+def read_file(fname, defaultonfailure=None, strip_end=True):
```
Is is bad practice to change global utility functions in a feature PR. Please remove this change from the PR. Also I guess this is no longer necessary as the configuration should be written as once and not read, modified and written back again.

zdc replied:
Removed this change. Most likely, this can lead to the unnecessary triggering of the `daemons` file updates (we have a newline in both templates and default files), but from the first view, this should not be a problem and may be fixed back in additional PR for `util.py`.

c-po commented on data/templates/frr/daemons.frr.tmpl:
```diff
@@ -0,0 +1,50 @@
+zebra=yes
```
Can we then also drop this file https://github.com/vyos/vyos-build/blob/current/data/live-build-config/hooks/live/30-frr-configs.chroot and call the new python script here (https://github.com/vyos/vyatta-cfg/blob/current/scripts/init/vyos-router#L305) on bootup same as we do with other parts to ensure a sane default config on bootup before loading the "real" config: https://github.com/vyos/vyatta-cfg/blob/current/scripts/init/vyos-router#L319-L324

zdc replied:
Yes, this would be the next logical step.

zdc committed:
FRR: T4020: Updated CLI options processing for FRR daemons

- Reverted changes from `python/vyos/util.py`. This may lead to
unnecessary FRR restart during each boot, depending on a default file
content and template, but makes this changeset cleaner.
- Fixed typos in node names (extra `>` characters).
- Added SNMP module for `isisd` and `ldpd`, since they have it compiled
now.

sever-sever commented on interface-definitions/system-frr.xml.in:
```diff
@@ -0,0 +1,66 @@
+          </leafNode>
+          <leafNode name="irdp">
+            <properties>
+              <help>>Enable ICMP Router Discovery Protocol support</help>
```
Unnecessary additional sign `>`

zdc replied:
Thanks! I did not notice them all.

sever-sever commented on interface-definitions/system-frr.xml.in:
```diff
@@ -0,0 +1,66 @@
+            <children>
+              <leafNode name="bgpd">
+                <properties>
+                  <help>>BGP</help>
```
Unnecessary additional sign `>`

sever-sever commented on interface-definitions/system-frr.xml.in:
```diff
@@ -0,0 +1,66 @@
+              </leafNode>
+              <leafNode name="ospf6d">
+                <properties>
+                  <help>>OSPFv3</help>
```
Unnecessary additional sign `>`

sever-sever commented on interface-definitions/system-frr.xml.in:
```diff
@@ -0,0 +1,66 @@
+              </leafNode>
+              <leafNode name="ospfd">
+                <properties>
+                  <help>>OSPFv2</help>
```
Unnecessary additional sign `>`

sever-sever commented on interface-definitions/system-frr.xml.in:
```diff
@@ -0,0 +1,66 @@
+              </leafNode>
+              <leafNode name="ripd">
+                <properties>
+                  <help>>RIP</help>
```
Unnecessary additional sign `>`

sever-sever commented on interface-definitions/system-frr.xml.in:
```diff
@@ -0,0 +1,66 @@
+              </leafNode>
+              <leafNode name="zebra">
+                <properties>
+                  <help>>Zebra (IP routing manager)</help>
```
Unnecessary additional sign `>`

c-po commented on python/vyos/util.py:
```diff
@@ -191,7 +191,6 @@ def read_file(fname, defaultonfailure=None):
         """ Read a file to string """
         with open(fname, 'r') as f:
             data = f.read().strip()
-        return data
```
Why no longer returning the data we have read? I guess this will break all kind of stuff.

zdc replied:
Wow... What a stupid mistake. Thanks, fixed.

andamasov commented:
@dmbaturin @c-po can you look please and merge
thanks!
